### PR TITLE
Mesh network

### LINKED ISSUE #28: Support multiple WebRTC connections

### LINKED ISSUE #29: Combine join and chat pages into one url
Instead of jumping between url's, organize the components onto a single page.

### PR DESCRIPTION
**Pull request by siberowl**
- closes #28 

- Also Closes #29 bc #33

### CODE REVIEW

**Comment by siberowl** [comment]
Thinking about removing all WebRTC logic from RoomPage.tsx.

**Comment by Njima1572** [comment]
> Thinking about removing all WebRTC logic from RoomPage.tsx.

Make it into a component or lib kind of removing??

File: client/src/RTCPeerConnector.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+			.catch((e) => {
+				console.warn(e);
+			});
+	});
+};
+
+// emit an answer when offer is received
+socket.on("OFFER", (dataString) => {
+	const sdp = JSON.parse(dataString).sdp;
+	const target = JSON.parse(dataString).id;
```
Comment (Njima1572) [inline client/src/RTCPeerConnector.ts:62]: Rename this to receiver_id

File: client/src/RTCPeerConnector.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			peerConnection
+				.createAnswer()
+				.then((answer) => {
+					return peerConnection.setLocalDescription(answer);
+				})
+				.then(() => {
+					const data = {
+						sdp: peerConnection.localDescription,
+						id: socket.id,
+						target: target,
```
Comment (Njima1572) [inline client/src/RTCPeerConnector.ts:77]: Rename to `sender_id` and `receiver_id`

File: client/src/RTCPeerConnector.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		})
+		.catch((e) => {
+			console.warn(e);
+		});
+});
+
+// set remote description once answer is recieved to establish connection
+socket.on("ANSWER", (dataString) => {
+	const sdp = JSON.parse(dataString).sdp;
+	const target = JSON.parse(dataString).id;
```
Comment (Njima1572) [inline client/src/RTCPeerConnector.ts:94]: If socket.id === target.id, then return or whatever to end the function

**Comment by Njima1572** [comment]
Server should hold the information of who is in the room already so that new user can get the list

File: client/src/RTCPeerConnector.ts (hunk 3 above)
Comment (siberowl) [reply client/src/RTCPeerConnector.ts]: Realized the server only broadcasts it (broadcast.to) to the receiver in the first place so no need to check it again in RTCPeerConnector.
https://github.com/Njima1572/spatialchatclone/blob/c7e9a070c6e142b5d23de45ddd76ebab00580ef9/server/src/index.ts#L38

File: client/src/RTCPeerConnector.ts (hunk 2 above)
Comment (siberowl) [reply client/src/RTCPeerConnector.ts]: fixed

**Comment by siberowl** [comment]
Calling initialization functions for the WebRTC network in RoomPage.tsx now. Still untested
https://github.com/Njima1572/spatialchatclone/blob/c7e9a070c6e142b5d23de45ddd76ebab00580ef9/client/src/pages/RoomPage.tsx#L11-L33

**Comment by siberowl** [comment]
There are still some bugs that need to be identified and fixed (to be addressed using cypress in #35 ), but the basic voice chatting flow between multiple users works.

File: server/src/index.ts [hunk 4]
```diff
@@ -6,34 +6,46 @@ import cors from "cors";
...
+	});
+	socket.on("OFFER", (dataString) => {
+		const targetId = JSON.parse(dataString).receiverId;
+		socket.broadcast.to(targetId).emit("OFFER", dataString);
+	});
+	socket.on("ANSWER", (dataString) => {
+		const targetId = JSON.parse(dataString).receiverId;
+		socket.broadcast.to(targetId).emit("ANSWER", dataString);
+	});
+	socket.on("disconnect", () => {
```
Comment (Njima1572) [inline server/src/index.ts:43]: Make this uppercase instead of lower case to match the other ones?

File: client/src/DeviceSelector.tsx [hunk 5]
```diff
@@ -1,61 +1,57 @@
-import React, { useState, useEffect } from "react";
```
Comment (Njima1572) [inline client/src/DeviceSelector.tsx:1]: Run Prettier on this file

File: client/src/User.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	peerConnection: any;
+	sender: RTCRtpSender;
+	id: string;
+	stream: MediaStream;
+	player: HTMLAudioElement;
+	constructor(id: string) {
+		this.id = id;
+		this.sender = null as unknown as RTCRtpSender;
+		this.peerConnection = new RTCPeerConnection({
+			iceServers: [{ urls: "stun:iphone-stun.strato-iphone.de:3478" }],
```
Comment (Njima1572) [inline client/src/User.ts:11]: https://www.3cx.com/pbx/what-is-a-stun-server/#:~:text=A%20STUN%20(Session%20Traversal%20of,outside%20of%20the%20local%20network.&text=The%20RFC%20states%20that%20this%20port%20and%20IP%20are%20arbitrary.

File: server/src/index.ts (hunk 4 above)
Comment (siberowl) [reply server/src/index.ts:43]: "connect" and "disconnect" are built in events so I don't think I can change them. Will consolidate the event names into a separate constants file and make them lowercase in the refactor branch.

File: client/src/User.ts (hunk 6 above)
Comment (siberowl) [reply client/src/User.ts:12]: Added comment for clarification.

**Pull request merged**